### ISSUE TITLE
Middleware Compatibility

### ISSUE THREAD

**Issue by mpeyper** [ioof-holdings/redux-subspace#38]
After the v2 improvements in #34 we are in a better position to support more existing redux middlewares.

This started with [redux-saga](https://redux-saga.js.org/) (which was added in #34), then [redux-promise](https://github.com/acdlite/redux-promise) (#36) and now [redux-observable](https://redux-observable.js.org/) (if #37 is merged).

My plan is to try out the most popular middleware and either document that they work, or list any caveats in the [redux-subspace documentation](https://github.com/ioof-holdings/redux-subspace/blob/master/docs/advanced/middleware/README.md) (like [redux-thunk](https://github.com/ioof-holdings/redux-subspace/blob/master/docs/advanced/middleware/redux-thunk.md) and [redux-promise](https://github.com/ioof-holdings/redux-subspace/blob/master/docs/advanced/middleware/redux-promise.md)), and, if required create a compatibility package for them (like [redux-saga](https://github.com/ioof-holdings/redux-subspace/blob/master/docs/advanced/middleware/redux-saga.md)).

To start with, I want to at least cover the middleware in the [ecosystem section of the Redux docs](http://redux.js.org/docs/introduction/Ecosystem.html#middleware), but also other's as they come on my radar, such as [redux-logic](https://github.com/jeffbski/redux-logic) which was requested in #30.

This would be a great task for others to get their hands dirty in redux-subspace.

**Comment by mpeyper** [comment]
## Middleware:

- [x] [redux-thunk](http://github.com/gaearon/redux-thunk) - [documented](https://github.com/ioof-holdings/redux-subspace/blob/master/docs/advanced/middleware/redux-thunk.md)
- [x] [redux-promise](https://github.com/acdlite/redux-promise) - [documented](https://github.com/ioof-holdings/redux-subspace/blob/master/docs/advanced/middleware/redux-promise.md)
- [] [redux-axios-middleware](https://github.com/svrcekmichal/redux-axios-middleware)
- [] [redux-observable](https://github.com/redux-observable/redux-observable/) - [compatability package PR](https://github.com/ioof-holdings/redux-subspace/pull/37)
- [] [redux-cycles](https://github.com/cyclejs-community/redux-cycles)
- [] [redux-logger](https://github.com/fcomb/redux-logger)
- [] [redux-immutable-state-invariant](https://github.com/leoasis/redux-immutable-state-invariant)
- [] [redux-unhandled-action](https://github.com/socialtables/redux-unhandled-action)
- [] [redux-analytics](https://github.com/markdalgleish/redux-analytics)
- [] [redux-gen](https://github.com/weo-edu/redux-gen)
- [x] [redux-saga](https://github.com/yelouafi/redux-saga) - [compatability package](https://github.com/ioof-holdings/redux-subspace/tree/master/packages/redux-subspace-saga)
- [] [redux-action-tree](https://github.com/cerebral/redux-action-tree)
- [] [apollo-client](https://github.com/apollostack/apollo-client)
- [] [redux-logic](https://github.com/jeffbski/redux-logic)

**Comment by mpeyper** [comment]
Updated list with some enhancers as well after the redux-loop addition in #44

**Comment by jcheroske** [comment]
Man, seeing redux-logic at the bottom of that list is painful. It's the one middleware to rule them all!

**Comment by mpeyper** [comment]
The order of that list is not indicative of the order they will be looked at... But if it makes you feel better, I'll move it up the list ;)

**Comment by jcheroske** [comment]
I don't want to bash other middlewares, but, having used several of them, I find `redux-logic` to be a kind of meta-middleware that rolls the functionality of the different ones together into a really nice, seamless whole. I usually write my side-effects using async/await, and it supports that syntax right out of the box. But if I need more power, then I can switch over to rxjs syntax with no pain, while still getting many of the other perks that `redux-logic` has to offer. I try to spread the word on it, simply because it deserves much more attention.

**Comment by mpeyper** [comment]
For me, there are pros and cons both ways.  It great when something is really flexible and can be used how you want to use it, but it can sometimes become a jack-of-all-trade, master-of-none.  Something that is super good at one thing is also great when you're doing that thing, but often falls down when you want to so something slightly left of normal.  Of course, it's possible to balance the line between being flexible enough to cover a greater number of cases, without sacrificing features in what you can do, but very few actually pull this off well, or have a massive amount of effort go into maintaining that balance.

In my personal usage of Redux, I actually haven't come across many times that redux-thunk wasn't enough for what I was trying to accomplish.

I look forward to trying redux-logic based on your glowing review.  Out of interest, have you tried redux-cycles yet? I've heard a lot of good things about it which is why it's so high on my list of middleware to look at.

**Comment by mpeyper** [comment]
FYI @jcheroske, I've begun working (in my spare time, so don't get too excited on it being done soon) on a redux-logic wrapper and I have to admit, it's a pretty sweet middleware.

My only complaint (which would literally only be an issue for me) is that there isn't one single place the logic exposes the store (or it's functions), meaning I have recreate the subspace multiple times (in `validate`, in `transform` and in `process` every time the logic is triggered), so it may add some performance overhead, but in my testing it hasn't been too bad.  I'm still coming to grips with how the logic objects work so I may yet find better way.

I do have a quick question on usage for you.  According to the [usage section of their docs](https://github.com/jeffbski/redux-logic#usage), it seems as though logics are combined through standard array combination:

```
// in logic.js - combines logic from across many files, just
// a simple array of logic to be used for this app
export default [
 ...todoLogic,
 ...pollsLogic
];
```

```
// pollsLogic
export default [
  validationLogic,
  addUniqueId,
  fetchPollsLogic
];
```

The common pattern with `redux-subspace` is to wrap each layer with a `subspaced` function, so which is preffered

```
// Option 1
export default [
 ...subspaced((state) => state.todos, 'todos')(todoLogic),
 ...subspaced((state) => state.polls, 'polls')(pollsLogic)
];
```

or

```
// Option 2
export default [
 subspaced((state) => state.todos, 'todos')(todoLogic),
 subspaced((state) => state.polls, 'polls')(pollsLogic)
];
```

Basically the only difference is where the spread operators go, but option 2 is probably going to be harder to implement (it will need to return a single logic that can run multiple logics).

With option 1, I think the basic logic (pun intended) would be if an array is provided, an array is returned, but if a single logic is provided a single logic is returned.

Any thoughts and feedback are welcome.

**Comment by mpeyper** [comment]
The [Redux Ecosystem](https://redux.js.org/docs/introduction/Ecosystem.html) docs have recently been updated (it include redux-subspace not to - YAY!) so I'll try to find some time soon to go through that and see if [the list](https://github.com/ioof-holdings/redux-subspace/issues/38#issuecomment-323878087) is still valid and add and/or remove any items that make sense to.

FYI, I started looking at redux-logic again recently and it's still giving me some grief.  If anyone has experience with it, I'd love some input.

**Comment by mpeyper** [comment]
I don't see us adding taking on any more official packages to support.  I'm more than happy to consider any though, but I'm not going to keep track of them here anymore.  Closing.
